File: legacy_cpp/src/parser/parser.cpp
#include "parser.h"
#include "../tokenizer/tokenizer.h"
#include <iostream>
#include <stdexcept>

Parser::Parser(const std::vector<Token> &tokens) : tokens(tokens), pos(0) {}

const Token &Parser::peek() const {
  if (is_eof())
    return tokens.back();
  return tokens[pos];
}

const Token &Parser::advance() {
  if (is_eof())
    return tokens.back();
  return tokens[pos++];
}

bool Parser::is_eof() const {
  return pos >= tokens.size() || tokens[pos].type == TokenType::END_OF_FILE;
}

bool Parser::match(TokenType type, const std::string &value) {
  if (is_eof())
    return false;
  const Token &t = peek();
  if (t.type == type && (value.empty() || t.value == value)) {
    advance();
    return true;
  }
  return false;
}

void Parser::expect(TokenType type, const std::string &value) {
  if (!match(type, value)) {
    std::string expected =
        value.empty() ? "token type " + std::to_string((int)type) : value;
    throw std::runtime_error("Syntax error: expected " + expected +
                             " but found " + peek().value);
  }
}

std::shared_ptr<Query> Parser::parse() {
  if (match(TokenType::KEYWORD, "CREATE")) {
    return parse_create_table();
  } else if (match(TokenType::KEYWORD, "INSERT")) {
    return parse_insert();
  } else if (match(TokenType::KEYWORD, "SELECT")) {
    return parse_select();
  } else {
    throw std::runtime_error("Syntax error: unexpected token " + peek().value);
  }
}

// CREATE TABLE <table_name> (<column1>, <column2>, ...)
std::shared_ptr<Query> Parser::parse_create_table() {
  auto query = std::make_shared<CreateTableQuery>();

  expect(TokenType::KEYWORD, "TABLE");

  const Token &name_token = advance();
  if (name_token.type != TokenType::IDENTIFIER) {
    throw std::runtime_error("Syntax error: expected table name");
  }
  query->table_name = name_token.value;

  expect(TokenType::SYMBOL, "(");

  // Parse columns
  while (!match(TokenType::SYMBOL, ")")) {
    const Token &col_token = advance();
    if (col_token.type != TokenType::IDENTIFIER) {
      throw std::runtime_error("Syntax error: expected column name");
    }
    query->columns.push_back(col_token.value);

    if (!match(TokenType::SYMBOL, ",")) {
      expect(TokenType::SYMBOL, ")");
      break;
    }
  }

  return query;
}

// INSERT INTO <table_name> VALUES (<value1>, <value2>, ...)
std::shared_ptr<Query> Parser::parse_insert() {
  auto query = std::make_shared<InsertQuery>();

  expect(TokenType::KEYWORD, "INTO");

  const Token &name_token = advance();
  if (name_token.type != TokenType::IDENTIFIER) {
    throw std::runtime_error("Syntax error: expected table name");
  }
  query->table_name = name_token.value;

  expect(TokenType::KEYWORD, "VALUES");
  expect(TokenType::SYMBOL, "(");

  // Parse values
  while (!match(TokenType::SYMBOL, ")")) {
    const Token &val_token = advance();
    if (val_token.type == TokenType::STRING ||
        val_token.type == TokenType::INTEGER) {
      query->values.push_back(val_token.value);
    } else {
      throw std::runtime_error(
          "Syntax error: expected value (string or integer)");
    }

    if (!match(TokenType::SYMBOL, ",")) {
      expect(TokenType::SYMBOL, ")");
      break;
    }
  }

  return query;
}

// SELECT <column_name> FROM <table_name> [WHERE <column_name> <operator>
// <value>]
std::shared_ptr<Query> Parser::parse_select() {
  auto query = std::make_shared<SelectQuery>();

  // Parse columns
  const Token &col_token = advance();
  if (col_token.type == TokenType::IDENTIFIER) {
    query->columns.push_back(col_token.value);
  } else if (col_token.type == TokenType::SYMBOL && col_token.value == "*") {
    query->columns.push_back("*");
  } else {
    throw std::runtime_error("Syntax error: expected column name");
  }

  expect(TokenType::KEYWORD, "FROM");

  const Token &name_token = advance();
  if (name_token.type != TokenType::IDENTIFIER) {
    throw std::runtime_error("Syntax error: expected table name");
  }
  query->table_name = name_token.value;

  if (match(TokenType::KEYWORD, "WHERE")) {
    query->has_where = true;

    const Token &where_col = advance();
    if (where_col.type != TokenType::IDENTIFIER)
      throw std::runtime_error("Syntax error: expected column in WHERE");
    query->where_clause.column = where_col.value;

    const Token &op_token = advance();
    if (op_token.type != TokenType::OPERATOR)
      throw std::runtime_error("Syntax error: expected operator in WHERE");
    query->where_clause.op = op_token.value;

    const Token &val_token = advance();
    if (val_token.type != TokenType::STRING &&
        val_token.type != TokenType::INTEGER) {
      throw std::runtime_error("Syntax error: expected value in WHERE");
    }
    query->where_clause.value = val_token.value;
  }

  return query;
}

File: src/main/java/com/coresql/parser/Parser.java
package com.coresql.parser;

import com.coresql.ast.*;
import com.coresql.tokenizer.Token;
import com.coresql.tokenizer.TokenType;

import java.util.List;

public class Parser {
    private final List<Token> tokens;
    private int pos = 0;

    public Parser(List<Token> tokens) {
        this.tokens = tokens;
    }

    private Token peek() {
        if (isEof()) return tokens.get(tokens.size() - 1);
        return tokens.get(pos);
    }

    private Token advance() {
        if (isEof()) return tokens.get(tokens.size() - 1);
        return tokens.get(pos++);
    }

    private boolean isEof() {
        return pos >= tokens.size() || tokens.get(pos).type() == TokenType.END_OF_FILE;
    }

    private boolean match(TokenType type, String value) {
        if (isEof()) return false;
        Token t = peek();
        if (t.type() == type && (value.isEmpty() || t.value().equals(value))) {
            advance();
            return true;
        }
        return false;
    }

    private boolean match(TokenType type) {
        return match(type, "");
    }

    private void expect(TokenType type, String value) {
        if (!match(type, value)) {
            String expected = value.isEmpty() ? "token type " + type : value;
            throw new IllegalArgumentException("Syntax error: expected " + expected + " but found " + peek().value());
        }
    }

    private void expect(TokenType type) {
        expect(type, "");
    }

    public Query parse() {
        Query query;
        if (match(TokenType.KEYWORD, "CREATE")) {
            query = parseCreateTable();
        } else if (match(TokenType.KEYWORD, "INSERT")) {
            query = parseInsert();
        } else if (match(TokenType.KEYWORD, "SELECT")) {
            query = parseSelect();
        } else {
            throw new IllegalArgumentException("Syntax error: unexpected token " + peek().value());
        }

        match(TokenType.SYMBOL, ";");

        if (!isEof()) {
            throw new IllegalArgumentException("Syntax error: trailing garbage at end of statement: " + peek().value());
        }

        return query;
    }

    private Query parseCreateTable() {
        CreateTableQuery query = new CreateTableQuery();
        
        expect(TokenType.KEYWORD, "TABLE");
        
        Token nameToken = advance();
        if (nameToken.type() != TokenType.IDENTIFIER) {
            throw new IllegalArgumentException("Syntax error: expected table name");
        }
        query.tableName = nameToken.value();
        
        expect(TokenType.SYMBOL, "(");
        
        // Parse columns
        while (!match(TokenType.SYMBOL, ")")) {
            Token colToken = advance();
            if (colToken.type() != TokenType.IDENTIFIER) {
                throw new IllegalArgumentException("Syntax error: expected column name");
            }

            Token typeToken = advance();
            if (typeToken.type() != TokenType.KEYWORD && typeToken.type() != TokenType.IDENTIFIER) {
                throw new IllegalArgumentException("Syntax error: expected column type");
            }

            query.columns.add(new ColumnDefinition(colToken.value(), typeToken.value()));
            
            if (!match(TokenType.SYMBOL, ",")) {
                expect(TokenType.SYMBOL, ")");
                break;
            }
        }
        
        return query;
    }

    private Query parseInsert() {
        InsertQuery query = new InsertQuery();
        
        expect(TokenType.KEYWORD, "INTO");
        
        Token nameToken = advance();
        if (nameToken.type() != TokenType.IDENTIFIER) {
            throw new IllegalArgumentException("Syntax error: expected table name");
        }
        query.tableName = nameToken.value();
        
        expect(TokenType.KEYWORD, "VALUES");
        expect(TokenType.SYMBOL, "(");
        
        // Parse values
        while (!match(TokenType.SYMBOL, ")")) {
            Token valToken = advance();
            if (valToken.type() == TokenType.STRING || valToken.type() == TokenType.INTEGER) {
                query.values.add(valToken.value());
            } else {
                throw new IllegalArgumentException("Syntax error: expected value (string or integer)");
            }
            
            if (!match(TokenType.SYMBOL, ",")) {
                expect(TokenType.SYMBOL, ")");
                break;
            }
        }
        
        return query;
    }

    private Query parseSelect() {
        SelectQuery query = new SelectQuery();
        
        // Parse columns
        Token colToken = advance();
        if (colToken.type() == TokenType.IDENTIFIER) {
            query.columns.add(colToken.value());
        } else if (colToken.type() == TokenType.SYMBOL && colToken.value().equals("*")) {
            query.columns.add("*"); 
        } else {
             throw new IllegalArgumentException("Syntax error: expected column name");
        }
        
        expect(TokenType.KEYWORD, "FROM");
        
        Token nameToken = advance();
        if (nameToken.type() != TokenType.IDENTIFIER) {
            throw new IllegalArgumentException("Syntax error: expected table name");
        }
        query.tableName = nameToken.value();
        
        if (match(TokenType.KEYWORD, "WHERE")) {
            query.hasWhere = true;
            query.whereClause = new Condition();
            
            Token whereCol = advance();
            if (whereCol.type() != TokenType.IDENTIFIER) throw new IllegalArgumentException("Syntax error: expected column in WHERE");
            query.whereClause.column = whereCol.value();
            
            Token opToken = advance();
            if (opToken.type() != TokenType.OPERATOR) throw new IllegalArgumentException("Syntax error: expected operator in WHERE");
            query.whereClause.op = opToken.value();
            
            Token valToken = advance();
            if (valToken.type() != TokenType.STRING && valToken.type() != TokenType.INTEGER) {
                throw new IllegalArgumentException("Syntax error: expected value in WHERE");
            }
            query.whereClause.value = valToken.value();
        }
        
        return query;
    }
}
